<a href="https://colab.research.google.com/github/comtechj/AIFFEL_quest_cr/blob/main/Main_Quest/0123_digits_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

digits = load_digits()

In [ ]:
digits_data = digits.data
print(digits_data.shape)

(1797, 64)


In [ ]:
digits_label = digits.target
print(digits_label.shape)

(1797,)


In [ ]:
digits_label_names = digits.target_names
print(digits_label_names)

[0 1 2 3 4 5 6 7 8 9]


In [ ]:
# digits.feature_names
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

:Number of Instances: 1797
:Number of Attributes: 64
:Attribute Information: 8x8 image of integer pixels in the range 0..16.
:Missing Attribute Values: None
:Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
:Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each block. This generates
an in

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(digits_data,
                                                    digits_label,
                                                    test_size = 0.2,
                                                    random_state = 15)

 ## Decision Tree 모델

In [ ]:
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state = 15)
decision_tree.fit(X_train, y_train)
tree_pred = decision_tree.predict(X_test)

# tree_pred == y_test
print(classification_report(tree_pred, y_test))

              precision    recall  f1-score   support

           0       0.90      1.00      0.95        28
           1       0.82      0.84      0.83        37
           2       0.87      0.73      0.80        45
           3       0.74      0.71      0.73        28
           4       0.80      0.94      0.87        35
           5       0.94      0.87      0.90        38
           6       0.89      0.87      0.88        39
           7       0.94      0.91      0.93        35
           8       0.77      0.79      0.78        34
           9       0.77      0.80      0.79        41

    accuracy                           0.84       360
   macro avg       0.85      0.85      0.84       360
weighted avg       0.85      0.84      0.84       360



 ## Random Forest 모델

In [ ]:
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state = 15)
random_forest.fit(X_train, y_train)
forest_pred = random_forest.predict(X_test)

# forest_pred == y_test
print(classification_report(forest_pred, y_test))

              precision    recall  f1-score   support

           0       0.94      0.97      0.95        30
           1       0.97      0.97      0.97        38
           2       1.00      1.00      1.00        38
           3       0.96      0.96      0.96        27
           4       0.98      0.95      0.96        42
           5       1.00      0.97      0.99        36
           6       0.97      1.00      0.99        37
           7       1.00      0.97      0.99        35
           8       0.94      0.97      0.96        34
           9       0.98      0.98      0.98        43

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.98      0.97      0.98       360



 ## SVM 모델

In [ ]:
from sklearn import svm

svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
svm_pred = svm_model.predict(X_test)

# svm_pred == y_test
print(classification_report(svm_pred, y_test))

              precision    recall  f1-score   support

           0       0.97      1.00      0.98        30
           1       1.00      0.95      0.97        40
           2       1.00      1.00      1.00        38
           3       0.96      0.96      0.96        27
           4       0.98      0.98      0.98        41
           5       1.00      1.00      1.00        35
           6       1.00      1.00      1.00        38
           7       1.00      1.00      1.00        34
           8       0.94      0.97      0.96        34
           9       0.98      0.98      0.98        43

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



 ## SGD Classifier 모델

In [ ]:
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
sgd_pred = sgd_model.predict(X_test)

# sgd_pred == y_test
print(classification_report(sgd_pred, y_test))

              precision    recall  f1-score   support

           0       0.97      1.00      0.98        30
           1       0.97      0.95      0.96        39
           2       0.95      1.00      0.97        36
           3       0.85      1.00      0.92        23
           4       0.98      1.00      0.99        40
           5       0.97      0.94      0.96        36
           6       0.95      1.00      0.97        36
           7       1.00      0.97      0.99        35
           8       0.94      0.80      0.87        41
           9       0.93      0.91      0.92        44

    accuracy                           0.95       360
   macro avg       0.95      0.96      0.95       360
weighted avg       0.95      0.95      0.95       360



 ## Logistic Regression 모델

In [ ]:
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression()

logistic_model.fit(X_train, y_train)
logistic_pred = logistic_model.predict(X_test)

# logistic_pred == y_test
print(classification_report(logistic_pred, y_test))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        31
           1       0.97      0.95      0.96        39
           2       0.97      1.00      0.99        37
           3       0.93      0.96      0.94        26
           4       1.00      0.98      0.99        42
           5       0.97      0.97      0.97        35
           6       0.97      1.00      0.99        37
           7       1.00      1.00      1.00        34
           8       0.97      0.92      0.94        37
           9       0.95      0.98      0.96        42

    accuracy                           0.97       360
   macro avg       0.97      0.98      0.97       360
weighted avg       0.98      0.97      0.97       360



/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


# **`SVM 모델이 98%로 정확도가 가장 높다.`**
**`대체적으로 정확도가 높다.`**